src/core/transaction.c

```diff
@@ -1107,20 +1121,27 @@ int transaction_add_isolate_jobs(Transaction *tr, Manager *m) {
         assert(m);
 
         HASHMAP_FOREACH_KEY(u, k, m->units) {
+                Unit *o;
 
-                /* ignore aliases */
+                /* Ignore aliases */
                 if (u->id != k)
                         continue;
 
-                if (u->ignore_on_isolate)
+                /* No need to stop inactive units */
+                if (UNIT_IS_INACTIVE_OR_FAILED(unit_active_state(u)) && !u->job)
                         continue;
 
-                /* No need to stop inactive jobs */
-                if (UNIT_IS_INACTIVE_OR_FAILED(unit_active_state(u)) && !u->job)
+                if (!shall_stop_on_isolate(tr, u))
                         continue;
 
-                /* Is there already something listed for this? */
-                if (hashmap_get(tr->jobs, u))
+                /* Keep units that are triggered by units we want to keep around. */
+                bool keep = false;
+                UNIT_FOREACH_DEPENDENCY(o, u, UNIT_ATOM_TRIGGERED_BY)
+                        if (!shall_stop_on_isolate(tr, o)) {
+                                keep = true;
```

Comment:
I think this is too simplistic. It only covers units that are one dependency away from a unit that we shall keep. But let's say that we have a unit A which triggers B which requires C. And A is part of the isolation target. When we visit C, we'll schedule a stop job for it.

Don't we need to do a more comprehensive walk where we keep marking units as excluded from isolation walking out further and further into the graph of dependencies?

Reply:
that sounds correct, but also risky - given this works and is at least a correct but partial fix, maybe it would be better to get this in 253 and then if needed rework later?

Reply:
I'm fine with that too.